[JENKINS-37062] incorporate changes from stapler 1.253 (servlet 3.1)

jtnord opened the pull request:
Downstream updates from stapler 1.253

The stapler API was built using servlet 2.5 and yet Jenkins uses 3.1.
This meant some of the code that stapler used was missing new methods
that where part of the updated spec and if a plugin happened to call
them you would end up with a LinkageError or some other crazyness.

This also make Jenkins depend on a dummy version of the old serlet-api
maven co-ordinates such that if any plugin gets aa dependency on it
transitivly (e.g. Jenkins test harness) they will get a version with no
code so the classpath should always be clean (this was more an issue for
eclipse than an mvn command)

See [JENKINS-37062](https://issues.jenkins-ci.org/browse/JENKINS-37062).

### Testing performed

* All unit tests ran - verified no *new* failures introduced.
* checked the war does not include the servlet-api.jar (or javax.servlet-api.jar)
* started Jenkins with java -jar jenkins.war and ran the install wizard

### Upstream changes

[Changelog](https://github.com/stapler/stapler/blob/master/CHANGELOG.md#1253) and [diffs](https://github.com/stapler/stapler/compare/stapler-parent-1.252...stapler-parent-1.253).

Also [Commons Codec changelog](https://commons.apache.org/proper/commons-codec/changes-report.html#a1.9).

### Proposed changelog entries

* Stapler library upgraded to 1.253:
  * Improved performance for some Blue Ocean operations.
  * Various changes which could be of interest to plugin developers, for which see component changelog.
* Commons Codec library upgraded to 1.9.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees
@jenkinsci/code-reviewers

jtnord committed:
[JENKINS-37062] incorporate changes from stapler servlet 3.1

Downstream updates from stapler/stapler#131

The stapler API was built using servlet 2.5 and yet Jenkins uses 3.1.
This meant some of the code that stapler used was missing new methods
that where part of the updated spec and if a plugin happened to call
them you would end up with a LinkageError or some other crazyness.

This also make Jenkins depend on a dummy version of the old serlet-api
maven co-ordinates such that if any plugin gets aa dependency on it
transitivly (e.g. Jenkins test harness) they will get a version with no
code so the classpath should always be clean (this was more an issue for
eclipse than an mvn command)

jtnord commented on pom.xml:
```diff
@@ -230,6 +230,19 @@ THE SOFTWARE.
+        <groupId>javax.servlet</groupId>
+        <!-- the old artifactID for the servlet API -->
+        <artifactId>servlet-api</artifactId>
+        <version>[0-SNAPSHOT]</version>
```
this needs releasing before this can be merged...  Not sure how to go about making that happen.
@daniel-beck is this just a case of a new PR to the repository permissions?

daniel-beck replied:
As I wrote on the dev list, yes.

jtnord replied:
released

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev commented on core/pom.xml:
```diff
@@ -472,9 +472,9 @@ THE SOFTWARE.
-      <groupId>javax.servlet</groupId>
-      <artifactId>jstl</artifactId>
-      <version>1.1.0</version>
+      <groupId>javax.servlet.jsp.jstl</groupId>
```
Any risk of compatibility issues in plugins?

jtnord replied:
there is always a risk - but 1.2 has been part of the Java EE spec since Java EE 5 (may 2006) that contained Servlet 2.5 so you likely got this at run time anyway.  Whilst 1.2.1 is not called out in any spec it is an old version (also of the 2006 era in the JCP)
Maybe we should ask the [spec lead](https://jcp.org/en/jsr/detail?id=52) who is now involved in the Jenkins community :-)
/cc @pelegri

daniel-beck commented:
I think a review by Maven maven @stephenc would be helpful here.

jtnord committed:
[JENKINS-37062] use release version of javax.servlet:servlet-api:0

jglick commented on pom.xml:
```diff
@@ -230,6 +230,19 @@ THE SOFTWARE.
+        <groupId>javax.servlet</groupId>
+        <!-- the old artifactID for the servlet API -->
+        <artifactId>servlet-api</artifactId>
+        <version>[0]</version>
```
What does this mean, if not `0`? A comment would be in order.

stephenc replied:
This is a hard version range dependency that can only be satisfied by one and only one version, `0`

When you do `<version>0</version>` you are telling Maven
> Use any version you like, but I think `0` would be a good idea

The "closest" suggestion wins.

As soon as you have some version ranges specified, then Maven starts collating the intersection set... so if we had `<version>[0,1.2)</version>` and a dependency had `<version>[1,1.3]</version>` then Maven will conclude that the selected version should be in the range `[1,1.2)`... now if the closest suggestion was `1.0.1` and that suggestion is valid for the required range... Maven *should* use that suggestion, but otherwise it just picks from the range (in a way that may be documented but I cannot recall)

Keep in mind, that `<optional>` dependencies are allowed to contribute to range requirements... they just do not force their way onto the classpath.

By specifying `[0]` we force this version over anything else. 
* If somebody has specified a version in the normal way... requirements always beat suggestions and `0` wins.
* If somebody has specified a version range... if there is no intersection their build will fail... if there is an intersection including `0` then that wins.

jglick replied:
OK…anyway, my comment stands: a comment in the source code explaining this would be appropriate.

jtnord replied:
@jglic comment left

jglick commented on pom.xml:
```diff
@@ -230,6 +230,19 @@ THE SOFTWARE.
         <version>${access-modifier-annotation.version}</version>
       </dependency>
+
+      <!--  make sure these plugins are never used by us or by any plugins which end up depening on this version -->
```
Plugins?

Also, typo.

jglick commented on war/pom.xml:
```diff
@@ -61,12 +61,8 @@ THE SOFTWARE.
         -->
         <exclusion>
-          <groupId>javax.servlet</groupId>
-          <artifactId>servlet-api</artifactId>
```
Do you not need to exclude `javax.servlet:javax.servlet-api`?

stephenc replied:
Nope... `<version>[0]</version>` globally excludes it for us!

jglick replied:
Doing some nonobvious tricks here with dependencies. Have you tried building a core snapshot and then selecting that as a `jenkins.version` from some plugin’s POM to verify that you get the right `mvn dependency:tree` and that `mvn hpi:run` works as expected and so on?

jtnord replied:
> Have you tried building a core snapshot and then selecting that as a jenkins.version from some plugin’s POM to verify that you get the right mvn dependency:tree and that mvn hpi:run works as expected and so on?

No but I have been doing that trick on an internal plugin so that it gets the correct servlet API.

I can give it a go..

jtnord replied:
in `jenkinsci/workflow-step-api-plugin.git`  (sha 8856adca44c476938eb18d62d5a8a5986f60334c)

```
mvn -Denforcer.skip -Djenkins.version=2.80-20171025.175859-1 -Djenkins-war.version=2.80-20171025.180130-1 clean install
mvn -Denforcer.skip -Djenkins.version=2.80-20171025.175859-1 -Djenkins-war.version=2.80-20171025.180130-1 hpi:run
```

install passes all tests and builds ok
hpi:run Jenkins starts up plugin is installed, created a pipeline job with `echo "hello"` build it verified the console output of the run.

enforcer was skipped as project was configured for java7 and obviously has bytecode for java8.
using java.level=8 was not OK as there where syntax errors in some of the generic bounds matching that made it past the java7 compiler

stephenc commented on pom.xml:
```diff
@@ -230,6 +230,19 @@ THE SOFTWARE.
+        <!-- the old artifactID for the servlet API -->
+        <artifactId>servlet-api</artifactId>
+        <version>[0]</version>
+        <scope>provided</scope>
```
IIRC This does not do what you think it does when in dependency management... I recommend removing the `<scope>` and `<optional>` from depMgmt

stephenc commented on pom.xml:
```diff
@@ -246,6 +259,12 @@ THE SOFTWARE.
+      <!-- make sure our dependency tree and all others are clean of the legacy servlet api.  -->
+      <groupId>javax.servlet</groupId>
+      <!-- the old artifactID for the servlet API -->
+      <artifactId>servlet-api</artifactId>
```
IIRC this will inherit the version as there is no `<scope>` element or `<optional>` element... but if there was a `<scope>` or a `<optional>` then this would not match your current depMgmt so the version would not be inferred... (though TBH this is an aspect of depMgmt that is poorly documented)

stephenc reviewed (commented):
Just a query on the way you have implemented depMgmt, if you have tested and you have confidence that it works, fine by me.

Otherwise 👍

jtnord reviewed (commented):
oop forgot to submit my responses - sorry for the delay.

jglick committed:
Update commons-codec to 1.9.

jglick commented on core/pom.xml:
```diff
@@ -39,7 +39,7 @@ THE SOFTWARE.
   <properties>
     <staplerFork>true</staplerFork>
-    <stapler.version>1.252</stapler.version>
+    <stapler.version>1.253-SNAPSHOT</stapler.version>
```
Now just use the release.

jglick reviewed (commented):
You will probably also need to port the `commons-codec` fix from #3103.

jtnord committed:
[JENKINS-37062] update to release version of stapler 1.253

jtnord committed:
[JENKINS-37062] update pom comment

servlet-api is not a plugin :)

jglick commented on core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java:
```diff
@@ -64,13 +70,13 @@ public EncryptedSlaveAgentJnlpFile(AccessControlled it, String viewName, String
         RequestDispatcher view = req.getView(it, viewName);
         if ("true".equals(req.getParameter("encrypt"))) {
-            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            final CapturingServletOutputStream csos = new CapturingServletOutputStream();
```
Or we could just remove this old hack and directly render the JNLP template into a memory buffer using raw Jelly calls or something, not using a mock response.

jtnord replied:
I went for the conservative approach here as the code was not super easy to understand what it was doing to begin with,  If someone wants to do a better thing then go for it!

jglick commented on core/src/main/java/jenkins/slaves/EncryptedSlaveAgentJnlpFile.java:
```diff
@@ -104,4 +110,52 @@ public void generateResponse(StaplerRequest req, StaplerResponse res, Object nod
+        public void setWriteListener(WriteListener writeListener) {
+            // we are always ready to write so we just call once to say we are ready.
+            try {
+                // should we do this on a separate thread to avoid deadlocaks?
```
typo

jtnord replied:
fixed

jglick commented on pom.xml:
```diff
@@ -230,6 +230,22 @@ THE SOFTWARE.
         <version>${access-modifier-annotation.version}</version>
       </dependency>
+      <dependency>
+        <!--  make sure these old servlet versions are never used by us or by any plugins which end up depening on this version -->
```
typo

jtnord replied:
fixed

jglick commented on pom.xml:
```diff
@@ -230,6 +230,22 @@ THE SOFTWARE.
       </dependency>
+      <dependency>
+        <!--  make sure these old servlet versions are never used by us or by any plugins which end up depening on this version -->
+        <!-- Jenkins test Harness tries to fudge this, causes issues with some IDEs -->
```
Is some matching PR to `jenkins-test-harness` needed here then?

jtnord replied:
JTH already uses servlet 3.x so that should be OK.

The plugin POM should have the hack for servlet 2.x vs 3.x removed but that will mean it will no longer work correctly with old cores (ones where we originally added 3.1 support) as plugin authors may end up with servlet 3.1 on the classpath for compilation but end up using deploying to a jenkins that uses 2.4

I'm ok with creating a new plugin-pom and removing support for any jenkins in the pre 3.1 servlet era.  (basically pre 2.0 (2016-04-20) )

jglick reviewed (approved):
A few questions but basically sounds fine.

jtnord committed:
[JENKINS-37062] add a comment for those not familiar with version ranges

Add a comment about what <version>[x]</version> means as not everyone is
familiar with its syntax.

jtnord committed:
Merge remote-tracking branch 'origin/master' into JENKINS-37062

jtnord committed:
[JENKINS-37062] fix some typos in comments and update others.

There was a few typos in comments that have been addressed.
Also update the comment about the servlet fudge in plugin-pom

jglick commented on pom.xml:
```diff
@@ -234,12 +234,21 @@ THE SOFTWARE.
-        <!--  make sure these old servlet versions are never used by us or by any plugins which end up depening on this version -->
-        <!-- Jenkins test Harness tries to fudge this, causes issues with some IDEs -->
+        <!--  make sure these old servlet versions are never used by us or by any plugins which end up depending on this version -->
+        <!--  plugin-pom tries to fudge servlet support to be compatible with cores < 2.0 and JTH which needs 3.x for jetty,
```
IIUC this comment will become misleading if & when https://github.com/jenkinsci/plugin-pom/pull/83 is merged and becomes established.

jglick commented on pom.xml:
```diff
@@ -234,12 +234,21 @@ THE SOFTWARE.
         <artifactId>servlet-api</artifactId>
         <version>[0]</version>
+        <!-- 
+              "[0]" is a range that must be exaclty 0
```
typo

jglick commented on pom.xml:
```diff
@@ -234,12 +234,21 @@ THE SOFTWARE.
+              this is different to "0" which is hint to use version 0.
+              therefore unless anyone else uses ranges (they should not) this version will always win
+              We have deployed a version 0 to jenkins repo which has an empty jar
+              This prevents conflicts between the old Servet API and the new Servlet API as the groupIDs have changed
```
typo

jtnord commented:
@reviewbybees done

oleg-nenashev reviewed (approved):
🐝 let's finally do something about that. Seems to be a bit YOLO, but it the worst case we can just pick previous weekly as LTS

oleg-nenashev commented:
@jglick recommends doing some manual testing and ATH before merging

jtnord commented:
@oleg-nenashev  see inline for the manual testing I have already performed.

> checked the war does not include the servlet-api.jar (or javax.servlet-api.jar)
 started Jenkins with java -jar jenkins.war and ran the install wizard

> hpi:run Jenkins starts up plugin is installed, created a pipeline job with echo "hello" build it verified the console output of the run.
enforcer was skipped as project was configured for java7 and obviously has bytecode for java8.
using java.level=8 was not OK as there where syntax errors in some of the generic bounds matching that made it past the java7 compiler

If someone wants to kick the ATH with this, that's fine but from my point of view this is ready to go.
alternatively merge it and if there are ATH failures ping me (but I want to get this into the weekly ASAP)

jglick commented:
The only other things I could think of that use unusual HTTP features and so are worth special manual testing (neither of which is covered by the ATH anyway AFAIK):

* HTTP-based CLI
* something using `ProgressiveRendering`, like `/asynchPeople`

jtnord commented:
@jglick verified both worked.

for the CLI command I used `java -jar "C:\Users\jnord\Downloads\jenkins-cli (1).jar" -s http://localhost:8080/ -http -noKeyAuth -auth admin:admin build myjob`
